training/CUQIpy-PyTorch/HMC.ipynb

```diff
@@ -0,0 +1,857 @@
+    "As mentioned in the introduction, Hamiltonian Monte Carlo (HMC) is a general but still very efficient sampler for sampling high-dimensional distributions. It is beyond the scope of this exercise to go into the details of HMC, but we instead give a short example showing how it compares to using the classical Metropolis-Hastings algorithm. For more details in the theory of HMC, we refer to the [original paper](https://arxiv.org/abs/1206.1901) by Neal and the [No-U-Turn Sampler](https://arxiv.org/abs/1111.4246) (NUTS) variant by Hoffman and Gelman.\n",
+    "\n",
+    "Suppose we were aiming to sample from 2 dimensional probability density function shaped like a donut. We could also have selected a higher dimensional example making Metropolis-Hastings look even worse, but it makes visualization more difficult.\n",
+    "\n",
+    "This example can be loaded from the `DistributionGallery` class in CUQIpy as follows."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "donut = cuqi.distribution.DistributionGallery('donut')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The donut distribution has a manually derived gradient for the logpdf. For illustration let us plot the pdf and arrows showing the gradient of the pdf at a few points.\n",
+    "\n",
+    "Note: Currently there is no `plot_pdf` method for CUQIpy distributions, so we must write our own plotting code. Also at this point we still use NumPy arrays instead of PyTorch tensors and the main CUQIpy package, but we will switch to PyTorch tensors and CUQIpy-PyTorch later."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def plot_pdf(dist, plot_grad=False):\n",
+    "    # Ranges for the plot\n",
+    "    m, n, nl, ng = 300, 300, 30, 30\n",
+    "    xmin, xmax, ymin, ymax = -4, 4, -4, 4\n",
+    "\n",
+    "    # evaluate PDF\n",
+    "    X, Y = np.meshgrid(np.linspace(xmin, xmax, m), np.linspace(ymin, ymax, n))\n",
+    "    Xf, Yf = X.flatten(), Y.flatten()\n",
+    "    pts = np.vstack([Xf, Yf]).T   # pts is (m*n, d)\n",
+    "    Z = donut.pdf(pts).reshape((m, n))\n",
+    "\n",
+    "    if plot_grad:\n",
+    "        # evaluate gradient\n",
+    "        Xg, Yg = np.meshgrid(np.linspace(xmin, xmax, ng), np.linspace(ymin, ymax, ng))\n",
+    "        Xfg, Yfg = Xg.flatten(), Yg.flatten()\n",
+    "        posg = np.vstack([Xfg, Yfg]).T  \n",
+    "        grad = donut.gradient(posg)\n",
+    "        norm = np.linalg.norm(grad, axis=0)\n",
+    "        u, v = grad[0, :]/norm, grad[1, :]/norm\n",
+    "\n",
+    "    # plot PDF and gradient\n",
+    "    plt.contourf(X, Y, Z, nl)\n",
+    "    plt.contour(X, Y, Z, nl, linewidths=0.5, colors='k') \n",
+    "    if plot_grad:\n",
+    "        plt.quiver(posg[:, 0], posg[:, 1], u, v, units='xy', scale=4, color='gray')\n",
+    "\n",
```

Comment:
Split cell here to separate function def and the call to generate the plot. Add a new markdown cell in between to simply say we now plot using the function defined.

Reply:
Good suggestion. Thanks!